src/symbolic_choice.ml

```diff
@@ -232,83 +125,88 @@ module WQ = struct
     Condition.broadcast q.cond;
     Mutex.unlock q.mutex
 
-  let with_produce q f ~started =
-    Mutex.lock q.mutex;
-    q.producers <- q.producers + 1;
-    Mutex.unlock q.mutex;
-    started ();
-    match f () with
-    | () ->
-      Mutex.lock q.mutex;
-      q.producers <- q.producers - 1;
-      if q.producers = 0 then Condition.broadcast q.cond;
-      Mutex.unlock q.mutex
-    | exception e ->
-      let bt = Printexc.get_raw_backtrace () in
-      fail q;
-      Printexc.raise_with_backtrace e bt
-
   let init () =
     { mutex = Mutex.create ()
     ; cond = Condition.create ()
     ; queue = Queue.create ()
-    ; producers = 0
+    ; pledges = 0
     ; failed = false
     }
 end
 
-module Counter = struct
-  type t =
-    { mutable c : int
-    ; mutex : Mutex.t
-    ; cond : Condition.t
-    }
-
-  let incr t =
-    Mutex.lock t.mutex;
-    t.c <- t.c + 1;
-    Condition.broadcast t.cond;
-    Mutex.unlock t.mutex
-
-  let wait_n t n =
-    Mutex.lock t.mutex;
-    while t.c < n do
-      Condition.wait t.cond t.mutex
-    done;
-    Mutex.unlock t.mutex
-
-  let create () = { c = 0; mutex = Mutex.create (); cond = Condition.create () }
-end
-
 module Multicore = struct
-  type 'a st = St of (Thread.t -> solver -> 'a * Thread.t) [@@unboxed]
+  module Prio = struct
+    type t = Default
 
-  type 'a t =
-    | Empty : 'a t
-    | Ret : 'a st -> 'a t
-    | Retv : 'a -> 'a t
-    | Bind : 'a t * ('a -> 'b t) -> 'b t
-    | Assert : S.vbool -> unit t
-    | Choice : S.vbool -> bool t
-    | Choice_i32 : S.int32 -> int32 t
-    | Trap : Trap.t -> 'a t
+    let default = Default
+  end
+
+  type prio = Prio.t
 
   type 'a eval =
     | EVal of 'a
     | ETrap of Trap.t
     | EAssert of Encoding.Expr.t
 
-  type 'a run_result = ('a eval * Thread.t) Seq.t
+  let bind_lift_eval x f g =
+    match x with
+    | EVal x -> f x
+    | EAssert a -> g (EAssert a)
+    | ETrap t -> g (ETrap t)
+
+  module Out = struct
+    type ('a, 'arg) t = 'arg -> ('a, 'arg) aux
+
+    and ('a, 'arg) aux =
+      | Now of 'a
+      | Yield of prio * ('a, 'arg) t
+      | Choice of (('a, 'arg) aux * ('a, 'arg) aux)
+      | Stop
+
+    let return x : _ t = fun _arg -> Now x
+
+    let rec bind_aux (x : _ aux) (f : _ -> _ aux) : _ aux =
+      match x with
+      | Now x -> f x
+      | Yield (prio, lx) -> Yield (prio, fun arg -> bind_aux (lx arg) f)
+      | Choice (mx1, mx2) -> Choice (bind_aux mx1 f, bind_aux mx2 f)
+      | Stop -> Stop
+
+    and bind (mx : ('a, 'arg) t) (f : 'a -> ('b, 'arg) t) : _ t =
```

Comment:
`let rec` instead of `and` should be enough ? It makes reasoning easier as it shows they are not mutually recursive.